pkg/cmd/cli/cmd/tag.go

```diff
@@ -0,0 +1,186 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	cmdutil "github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	"github.com/openshift/origin/pkg/image/registry/imagestreamimage"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+const (
+	tagLong = `Tag existing images into image streams
+
+The tag command allows you to take an existing tag or image from an image
+stream, or a Docker image pull spec, and set it as the most recent image for a
+tag in 1 or more other image streams. It is similar to the 'docker tag'
+command, but it operates on image streams instead.`
+
+	tagExample = `  // Tag the current image for the image stream 'openshift/ruby' and tag '2.0' into the image stream 'yourproject/ruby with tag 'tip':
+  $ %[1]s tag openshift/ruby:2.0 yourproject/ruby:tip
+
+  // Tag a specific image:
+  $ %[1]s tag openshift/ruby@sha256:6b646fa6bf5e5e4c7fa41056c27910e679c03ebe7f93e361e6515a9da7e258cc yourproject/ruby:tip
+
+  // Tag an external Docker image:
+  $ %[1]s tag --source=docker openshift/origin:latest yourproject/ruby:tip`
+)
+
+// NewCmdTag implements the OpenShift cli tag command.
+func NewCmdTag(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	var sourceKind string
+
+	cmd := &cobra.Command{
+		Use:     "tag [--source=SOURCETYPE] SOURCE DEST [DEST ...]",
+		Short:   "Tag existing images into image streams",
+		Long:    tagLong,
+		Example: fmt.Sprintf(tagExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunTag(f, out, cmd, args, sourceKind)
+			cmdutil.CheckErr(err)
+		},
+	}
+
+	cmd.Flags().StringVar(&sourceKind, "source", sourceKind, "Optional hint for the source type; valid values are 'imagestreamtag', 'istag', 'imagestreamimage', 'isimage', and 'docker'")
+
+	return cmd
+}
+
+func determineSourceKind(input string) string {
+	switch strings.ToLower(input) {
+	case "istag", "imagestreamtag":
```

Comment:
Add something to rest mapper or just have an else block in here after you ask it.